src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,79 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Currency;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private long maxId;
+    private List<Product> products;
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public synchronized Product getProduct(Long id) throws ProductNotFoundException {
+        return products.stream()
+                .filter(product -> id.equals(product.getId()))
+                .findAny()
+                .orElseThrow(ProductNotFoundException::new);
     }
 
     @Override
-    public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+    public synchronized List<Product> findProducts() {
+        return products.stream()
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public synchronized void save(Product product) {
```

Comment:
What if null passed to this method? Please consider adding null handling here and in other methods where needed

Reply:
How it is better to handle this situation? I imagine two solutions:
1. custom exception;  
2. if null passed:
- void function: no action;
- value-returning function: return null as value.

Reply:
Both approaches can be good. I would choose the one that make it easier for me to code later

Reply:
Fixed